### PR TITLE
feat: add variable overlap tiling strategy

### PR DESCRIPTION
**Pull request by edparris**
This PR contains an implementation of the VariableOverlapTilingStrategy that has been requested by some users with CV models who need to receive full tiles of real pixels. This new region and tile decomposition strategy has been implemented using the new TilingStrategy abstraction which will allow us to easily enable this behavior with minimal code changes.

**Figure 1:** An example 25Kx12K image with 4096 tiles, a minimum of 100 pixel overlap and a 10K region size divided into regions using the baseline tiling strategy (VariableTileTilingStrategy). Note that the region size and overlap are fixed at the requested values but that tiles at the edges of each region and the image are truncated.
image

**Figure 2:** The same example image divided into regions using the new VariableOverlapTilingStrategy. Note that the overlap distance and region boundaries have been adjusted to make cutting full 4096x4096 tiles from each region possible.
image

### Checklist

Before you submit a pull request, please make sure you have the following:
- [x] Code changes are compact and well-structured to facilitate easy review
- [x] Changes are documented in the README.md and other relevant documentation pages
- [x] PR title and description accurately reflect the changes and are detailed enough for historical tracking
- [x] PR contains tests that cover all new code and the code has been manual tested
- [x] All new dependencies are declared (if any), and no unnecessary libraries are added
- [x] Performance impacts (if any) of the changes are evaluated and documented
- [x] Security implications of the changes (if any) are reviewed and addressed
- [x] I have read the [Contributing Guidelines](../CONTRIBUTING.md) and agree to follow the [Code of Conduct](../CODE_OF_CONDUCT.md)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/aws/osml/model_runner/tile_worker/variable_overlap_tiling_strategy.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        :param overlap: the amount of overlap (w, h)
+
+        :return: a collection of region boundaries
+        """
+        logger.debug(
+            f"ENTER FixedTileTilingStrategy.compute_regions: {processing_bounds}, {region_size}, {tile_size}, {overlap}"
+        )
+
+        adjusted_overlap = self._calculate_overlap_for_full_tiles(processing_bounds[1], tile_size, overlap)
+        logger.debug(f"FixedTileTilingStrategy.compute_regions: adjusted_overlap = {adjusted_overlap}")
```
Comment (jtblack-aws) [inline src/aws/osml/model_runner/tile_worker/variable_overlap_tiling_strategy.py:39]: VariableOverlapTilingStrategy

Comment (jtblack-aws) [reply src/aws/osml/model_runner/tile_worker/variable_overlap_tiling_strategy.py:39]: also in other logging statements but won't mark each one individually.

File: src/aws/osml/model_runner/tile_worker/variable_overlap_tiling_strategy.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+        :param processing_bounds: the bounds of the full image or area of interest in pixels ((r, c), (w, h))
+        :param region_size: the size of the regions in pixels (w, h)
+        :param tile_size: the size of the tiles in pixels (w, y)
+        :param overlap: the amount of overlap (w, h)
+
+        :return: a collection of region boundaries
+        """
+        logger.debug(
+            f"ENTER FixedTileTilingStrategy.compute_regions: {processing_bounds}, {region_size}, {tile_size}, {overlap}"
```
[the remark this answers is not in the document]
Comment (edparris) [reply src/aws/osml/model_runner/tile_worker/variable_overlap_tiling_strategy.py]: fixed

File: src/aws/osml/model_runner/tile_worker/variable_overlap_tiling_strategy.py (hunk 1 above)
Comment (edparris) [reply src/aws/osml/model_runner/tile_worker/variable_overlap_tiling_strategy.py]: fixed

**Review by jtblack-aws — APPROVED** [review]

**Pull request merged**